src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala

```diff
@@ -528,6 +534,162 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
       Nil)
   }
 
+  def withConcurrentCompactionTestSetup(toggle: Boolean)(closure: String => Unit): Unit = {
+    // in order to slow down the merge operation and increase the conflict probability,
+    // we use a UDF with sleep as the merge condition
+    spark.udf.register("slowMergeCondition", (condition: Boolean) => {
+      Thread.sleep(10)
+      condition
+    })
+
+    withTempDir { dir =>
+      withSQLConf(DeltaSQLConf.CONCURRENT_COMPACTION_CONFLICT_DETECTION.key -> toggle.toString,
+                  // we enable merge repartition before write to slow down the merge operation,
+                  // thus increasing conflict probability
+                  DeltaSQLConf.MERGE_REPARTITION_BEFORE_WRITE.key -> "true") {
+        closure(dir.getAbsolutePath)
+      }
+    }
+  }
+
+  val concurrentCompactionTestPartitions = 3
+  val concurrentCompactionTestRangeSize = 10
+  val concurrentCompactionTestNumIterations = 10
+
+  def createConcurrentCompactionTestTable(location: String): Unit = {
+    spark.range(0)
```

Comment:
What's the purpose of this `createConcurrentCompactionTestTable` in general and `spark.range(0)` in particular?

Reply:
This method creates the empty test table to be used in the concurrent compaction test, so that the test iterations could be made simpler, as they can all assume the table already exists and simply perform their merge/compaction operations.

The `spark.range(0)` is just my favorite way of creating an empty DataFrame without having to explicitly specify the schema (it's a single integer column named `id`). Then I just add the `par` column in the next line with the desired number of partitions for the test.